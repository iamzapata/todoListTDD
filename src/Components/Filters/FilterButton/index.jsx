import React, { PureComponent } from "react"
import { func, string, node, bool } from "prop-types"

class FilterButton extends PureComponent {
  render() {
    const {
      filter,
      setFilter,
      isListEmpty,
      children,
      currentFilter
    } = this.props
    return (
      <button
        onClick={() => setFilter(filter)}
        disabled={isListEmpty || filter === currentFilter}
      >
        {children}
      </button>
    )
  }
}

FilterButton.propTypes = {
  filter: string.isRequired,
  setFilter: func.isRequired,
  children: node.isRequired,
  isListEmpty: bool.isRequired,
  currentFilter: string.isRequired
}

export default FilterButton
