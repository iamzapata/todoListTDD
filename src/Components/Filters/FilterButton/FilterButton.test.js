import React from "react"
import { render, cleanup, fireEvent } from "react-testing-library"
import "jest-dom/extend-expect"
import FilterButton from "./index"

describe("<FilterButton />", () => {
  afterEach(cleanup)

  const createProps = () => ({
    filter: "SHOW_ALL",
    setFilter: jest.fn(),
    isListEmpty: false,
    currentFilter: "SHOW_ALL"
  })

  it("Should render self", () => {
    const { container } = render(
      <FilterButton {...createProps()}>All</FilterButton>
    )

    expect(container.firstChild).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })

  it("Should render children correctly", () => {
    const { queryByText } = render(
      <FilterButton {...createProps()}>All</FilterButton>
    )
    expect(queryByText("All").firstChild).toBeTruthy()
  })

  it("Should call setFiler when clicked with correct filter argument", () => {
    const props = {
      ...createProps(),
      currentFilter: "SHOW_COMPLETED"
    }
    const { container } = render(<FilterButton {...props}>All</FilterButton>)

    const button = container.querySelector("button")

    fireEvent.click(button)

    expect(props.setFilter).toHaveBeenCalledTimes(1)
    expect(props.setFilter).toHaveBeenCalledWith("SHOW_ALL")
  })

  it("Should be disabled if todos list is empty", () => {
    const props = {
      ...createProps(),
      isListEmpty: true
    }
    const { container } = render(<FilterButton {...props}>All</FilterButton>)

    const button = container.querySelector("button")

    expect(button).toHaveAttribute("disabled")
  })

  it("Should be disabled if current button's filter is active", () => {
    const { container } = render(
      <FilterButton {...createProps()}>Completed</FilterButton>
    )

    const button = container.querySelector("button")

    expect(button).toHaveAttribute("disabled")
  })
})
