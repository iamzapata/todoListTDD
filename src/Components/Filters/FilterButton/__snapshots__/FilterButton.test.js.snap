// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterButton /> Should render self 1`] = `
<button
  disabled=""
>
  All
</button>
`;
